utils/data_loader.py

```diff
@@ -215,92 +219,31 @@ def get_user_examples(conn: sqlite3.Connection, user_id: str,
-        edges_list.append(e)
-
-    # Last split
-    d, e = process_df(rows[i:])
-    df_list.append(d)
-    edges_list.append(e)
-
-    data_objs = [create_data_obj(df, edges, y=torch.tensor(label), mode=mode) for df, edges in zip(df_list, edges_list)]
-    return data_objs
 
-
-def get_negative_examples(conn: sqlite3.Connection, user_id: str, num_examples: int,
-                          rows_per_example=200, mode=LoadMode.DROP) -> List[Data]:
-    """
-    Loads and processes data from a database to generate a list of PyTorch Geometric `Data` objects.
-    :param conn: Database connection
-    :param user_id: user_id of positive examples
-    :param mode: Mode for processing node attributes
-    :param rows_per_example: Number of rows to include in one example
-    :param num_examples: Number of negative examples to include
-    :return: List[torch_geometric.data.Data]:
-        A list of `Data` objects, where each object represents a processed examples containing:
-            - `x`: Node attributes as a tensor
-            - `edge_index`: Edge indices tensor of shape [2, num_edges].
-            - `y`: tensor(0)
-    """
-    cursor = conn.cursor()
-    cursor.execute("""
-        SELECT * FROM key_press
-        WHERE user_id != ?
-        ORDER BY press_time
-    """, (user_id,))
-    rows = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
-
-    total_rows = len(rows)
-    starting_indices = set()
-    index = 0
-    while index <= total_rows - rows_per_example:
-        if rows.iloc[index]['user_id'] == rows.iloc[index + rows_per_example - 1]['user_id'] and \
-                rows.iloc[index]['date'] == rows.iloc[index + rows_per_example - 1]['date']:
-                starting_indices.add(index)
-                index += rows_per_example
-        else:
-            index += 1
-
-    df_list = []
-    edges_list = []
-    total_starting_indices = len(starting_indices)
-    if num_examples >= total_starting_indices or num_examples == 0:
-        if num_examples > total_starting_indices:
-            print(f"Warning: expected {num_examples} examples, but got {total_starting_indices}")
-        for starting_index in starting_indices:
-            d, e = process_df(rows[starting_index:starting_index + rows_per_example])
-            df_list.append(d)
-            edges_list.append(e)
-    else:
-        selected_indices = random.sample(sorted(starting_indices), num_examples)
-        for starting_index in selected_indices:
-            d, e = process_df(rows[starting_index:starting_index + rows_per_example])
+    i = 0
+    while i + rows_per_example <= len(rows):
+        if rows.iloc[i]['timestamp'] == rows.iloc[i + rows_per_example - 1]['timestamp']:
+            # print(f"{user_id} [{i} : {i + rows_per_example}]")
+            d, e = process_df(rows.iloc[i : i + rows_per_example])
             df_list.append(d)
             edges_list.append(e)
+            i += offset
+        else:
```

Comment:
This else guarantees that all the data is from the same "writing session" right ?
Otherwise LGTM

Reply:
Yes, i+=1 guarantees that next example starts where it should